BaragonAgentService/src/main/java/com/hubspot/baragon/agent/handlebars/ResolveHostnameHelper.java

```diff
@@ -1,49 +1,42 @@
 package com.hubspot.baragon.agent.handlebars;
 
-import java.net.InetAddress;
-import java.net.UnknownHostException;
-import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import com.github.jknack.handlebars.Helper;
 import com.github.jknack.handlebars.Options;
-import com.google.common.cache.Cache;
-import com.google.common.cache.CacheBuilder;
-import com.google.common.net.HostAndPort;
+import com.hubspot.baragon.models.UpstreamInfo;
+import com.hubspot.baragon.utils.UpstreamResolver;
 
-public class ResolveHostnameHelper implements Helper<String> {
+public class ResolveHostnameHelper implements Helper<Object> {
+  private static final Logger LOG = LoggerFactory.getLogger(ResolveHostnameHelper.class);
 
   public static final String NAME = "resolveHostname";
 
-  private final Cache<String, String> resolveCache;
+  private final UpstreamResolver resolver;
 
-  public ResolveHostnameHelper(long maxSize, long expireAfterDays) {
-    this.resolveCache = CacheBuilder.newBuilder()
-        .maximumSize(maxSize)
-        .expireAfterAccess(expireAfterDays, TimeUnit.DAYS)
-        .build();
+  public ResolveHostnameHelper(UpstreamResolver resolver) {
+    this.resolver = resolver;
   }
 
   @Override
-  public CharSequence apply(String address, Options options) {
-    try {
-      if (address.contains(":")) {
-        HostAndPort hostAndPort = HostAndPort.fromString(address);
-        String cached = resolveCache.getIfPresent(hostAndPort.getHost());
-        String ip;
-        int port = hostAndPort.getPort();
-        if (cached != null) {
-          ip = cached;
-        } else {
-          ip = InetAddress.getByName(hostAndPort.getHost()).getHostAddress();
-          resolveCache.put(hostAndPort.getHost(), ip);
-        }
-        return String.format("%s:%d", ip, port);
-      } else {
-        return InetAddress.getByName(address).getHostAddress();
-      }
-    } catch (UnknownHostException uhe) {
-      // Don't let this exception block rendering of the template, the lb config check will still fail if the host is truly unknown
-      return address;
+  public CharSequence apply(Object upstream, Options options) {
+    if (upstream instanceof String) {
+      String address = ((String) upstream);
+      LOG.trace("Trying to resolve a String upstream of {}", address);
+      String resolved = resolver.resolveUpstreamDNS(address).or(address);
+      LOG.trace("Resolved {} to {}", address, resolved);
+      return resolved;
+    } else if (upstream instanceof UpstreamInfo) {
+      UpstreamInfo upstreamInfo = ((UpstreamInfo) upstream);
+      LOG.trace("Trying to resolve an UpstreamInfo upstream of {} with upstreamInfo.getResolvedUpstream() = {}, resolver.resolveUpstreamDNS(upstreamInfo.getUpstream()) = {}, upstreamInfo.getUpstream() = {}", upstreamInfo, upstreamInfo
+          .getResolvedUpstream(), resolver.resolveUpstreamDNS(upstreamInfo.getUpstream()), upstreamInfo.getUpstream());
```

Comment:
Looks like this trace line will cause us to call resolveUpstreamDNS twice rather than one and storing in a variable for logging.

Reply:
Good catch